packages/blocks/src/page-block/edgeless/components/component-toolbar/component-toolbar.ts

```diff
@@ -220,33 +212,83 @@ export class EdgelessComponentToolbar extends WithDisposable(LitElement) {
     ]).forEach(slot =>
       this._disposables.add(slot.on(this._updateOnSelectedChange))
     );
+
+    this.style.position = 'absolute';
+    this.style.zIndex = '1';
+  }
+
+  protected override firstUpdated() {
+    const { _disposables } = this;
+    _disposables.add(
+      this.surface.viewport.slots.viewportUpdated.on(() =>
+        this._updatePosition()
+      )
+    );
   }
 
-  private _getCreateGroupButton() {
+  private _CreateGroupButton() {
     return html`<edgeless-add-group-button
       .edgeless=${this.edgeless}
     ></edgeless-add-group-button> `;
   }
 
-  private _getCreateFrameButton() {
+  private _CreateFrameButton() {
     return html`<edgeless-add-frame-button
       .edgeless=${this.edgeless}
     ></edgeless-add-frame-button>`;
   }
 
-  private _getReleaseFromGroupButton() {
+  private _ReleaseFromGroupButton() {
     return html`<edgeless-release-from-group-button
       .surface=${this.surface}
     ></edgeless-release-from-group-button>`;
   }
 
-  private _getAlignButton() {
+  private _AlignButton() {
     return html`<edgeless-align-button
       .edgeless=${this.edgeless}
     ></edgeless-align-button>`;
   }
 
+  private _Divider() {
+    return html`<component-toolbar-menu-divider></component-toolbar-menu-divider>`;
+  }
+
+  private async _updatePosition() {
+    // await this.updateComplete;
+    const { selectionManager } = this.edgeless;
+
+    const bound = edgelessElementsBound(selectionManager.elements);
+
+    const { viewport } = this.edgeless.surface;
+    const { width, height } = viewport;
+    const [x, y] = viewport.toViewCoord(bound.x, bound.y);
+
+    const [right, bottom] = viewport.toViewCoord(bound.maxX, bound.maxY);
+    const rect = this.getBoundingClientRect();
+
+    if (x >= width || right <= 0 || y >= height || bottom <= 0) {
+      this.style.left = (right <= 0 ? x - rect.width : x) + 'px';
+      this.style.top = (bottom <= 0 ? y - rect.height : y) + 'px';
+      return;
+    }
+
+    let offset = 34;
+    if (this.selection.elements.find(ele => isFrameBlock(ele))) {
+      offset += 10;
+    }
+    let top = y - rect.height - offset;
+    top < 0 && (top = y + bound.h * viewport.zoom + offset);
+
+    const left = clamp(x, 10, width - rect.width - 10);
+    top = clamp(top, 10, height - rect.height - 100);
+
+    this.style.left = left + 'px';
+    this.style.top = top + 'px';
+  }
+
   override render() {
+    this._updatePosition();
```

Comment:
I think it's anti-pattern to perform this kind of side effect in render function.